Rename EnabledIf to Validator and allow it to be specifed together with EnabledBy

szewai opened the pull request:
#### 865a57c2c9e5255556bb9c2ffb620e24d00907b4
<pre>
Rename EnabledIf to Validator and allow it to be specifed together with EnabledBy
<a href="https://bugs.webkit.org/show_bug.cgi?id=283471">https://bugs.webkit.org/show_bug.cgi?id=283471</a>
<a href="https://rdar.apple.com/140327105">rdar://140327105</a>

Reviewed by Timothy Hatcher.

In existing implementation, EnabledIf and EnabledBy are both recognized as &quot;runtime enablement conditon&quot;, and they
cannot co-exist. However, they are a bit different: EnabledBy should be specifically used for feature flag, and will be
enforced later on new message endpoints for new features; while EnabledIf can be used for more complicated condition,
like validating the receiver is in an active state. To distinguish between them, rename EnabledIf to Validator. Also,
allow message endpoints to be annotated with both of them since the purpose can be different.

* Source/WebKit/Scripts/webkit/messages.py:
(handler_function):
(generate_runtime_enablement):
(async_message_statement):
* Source/WebKit/Scripts/webkit/messages_unittest.py:
* Source/WebKit/Scripts/webkit/model.py:
(Message.__init__):
* Source/WebKit/Scripts/webkit/parser.py:
(parse):
* Source/WebKit/Scripts/webkit/tests/Makefile:
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::jsValueForArguments):
(IPC::jsValueForReplyArguments):
(IPC::messageArgumentDescriptions):
(IPC::messageReplyArgumentDescriptions):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessageReceiver.cpp:
(WebKit::TestWithEnabledBy::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessageReceiver.cpp: Removed.
* Source/WebKit/Scripts/webkit/tests/TestWithValidator.messages.in: Renamed from Source/WebKit/Scripts/webkit/tests/TestWithEnabledIf.messages.in.
* Source/WebKit/Scripts/webkit/tests/TestWithValidatorMessageReceiver.cpp: Added.
(WebKit::TestWithValidator::didReceiveMessage):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithValidator_AlwaysEnabled&gt;):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithValidator_EnabledIfPassValidation&gt;):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithValidator_EnabledIfSomeFeatureEnabledAndPassValidation&gt;):
* Source/WebKit/Scripts/webkit/tests/TestWithValidatorMessages.h: Renamed from Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessages.h.
(Messages::TestWithValidator::messageReceiverName):
(Messages::TestWithValidator::AlwaysEnabled::name):
(Messages::TestWithValidator::AlwaysEnabled::AlwaysEnabled):
(Messages::TestWithValidator::AlwaysEnabled::arguments):
(Messages::TestWithValidator::EnabledIfPassValidation::name):
(Messages::TestWithValidator::EnabledIfPassValidation::EnabledIfPassValidation):
(Messages::TestWithValidator::EnabledIfPassValidation::arguments):
(Messages::TestWithValidator::EnabledIfSomeFeatureEnabledAndPassValidation::name):
(Messages::TestWithValidator::EnabledIfSomeFeatureEnabledAndPassValidation::EnabledIfSomeFeatureEnabledAndPassValidation):
(Messages::TestWithValidator::EnabledIfSomeFeatureEnabledAndPassValidation::arguments):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.messages.in:

Canonical link: <a href="https://commits.webkit.org/287014@main">https://commits.webkit.org/287014@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d1d323685c7bbea647e6b824f83f6cce1ded739a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77927 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56960 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31302 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82575 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29184 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66124 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5255 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61020 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18946 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80995 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51011 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66992 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41322 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/77436 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48374 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27527 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/71071 "Found 5 new JSC stress test failures: stress/get-by-val-hoist-above-structure-2.js.ftl-no-cjit-b3o0, stress/get-by-val-hoist-above-structure.js.ftl-eager-no-cjit-b3o1, stress/get-by-val-hoist-above-structure.js.ftl-no-cjit-no-put-stack-validate, wasm.yaml/wasm/modules/wasm-imports-js-exports.js.wasm-no-cjit, wasm.yaml/wasm/v8/regress/regress-824681.js.wasm-collect-continuously (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69468 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24758 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83937 "Built successfully") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/77162 "Found 3 new JSC stress test failures: stress/attribute-custom-accessor.js.dfg-eager, stress/get-by-val-hoist-above-structure-2.js.ftl-eager-no-cjit-b3o1, wasm.yaml/wasm/v8/regress/regress-824681.js.wasm-collect-continuously (failure)") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5294 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3569 "Found 2 new test failures: ipc/create-media-source-with-invalid-constraints-crash.html webrtc/vp8-then-h264.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69238 "Found 1 new test failure: compositing/shared-backing/overlap-after-end-sharing.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5450 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66836 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68494 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17112 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12495 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10621 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/99475 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5242 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7995 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21723 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5234 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8666 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7019 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon reviewed (changes requested):
We should not merge this. We should be moving old code off of `MESSAGE_CHECK` instead of removing `EnabledIf`.

xeenon commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
```diff
@@ -131,8 +131,8 @@ messages -> NetworkConnectionToWebProcess WantsDispatchMessage {
-    [EnabledIf='isWebTransportEnabled()'] InitializeWebTransportSession(URL url) -> (std::optional<WebKit::WebTransportSessionIdentifier> identifier)
-    [EnabledIf='isWebTransportEnabled()'] DestroyWebTransportSession(WebKit::WebTransportSessionIdentifier identifier)
+    [Validator=isWebTransportEnabled] InitializeWebTransportSession(URL url) -> (std::optional<WebKit::WebTransportSessionIdentifier> identifier)
+    [Validator=isWebTransportEnabled] DestroyWebTransportSession(WebKit::WebTransportSessionIdentifier identifier)
```
You should leave this as-is, since you are changing it in another PR to EnabledBy. EnabledBy makes sense here.

szewai replied:
Yes, I am just changing them here so this patch could build and run test; I am going to drop these now that the other patch is merged. Thanks for review!

szewai committed:
Rename EnabledIf to Validator and allow it to be specifed together with EnabledBy
https://bugs.webkit.org/show_bug.cgi?id=283471
rdar://140327105

Reviewed by Timothy Hatcher.

In existing implementation, EnabledIf and EnabledBy are both recognized as "runtime enablement conditon", and they
cannot co-exist. However, they are a bit different: EnabledBy should be specifically used for feature flag, and will be
enforced later on new message endpoints for new features; while EnabledIf can be used for more complicated condition,
like validating the receiver is in an active state. To distinguish between them, rename EnabledIf to Validator. Also,
allow message endpoints to be annotated with both of them since the purpose can be different.

* Source/WebKit/Scripts/webkit/messages.py:
(handler_function):
(generate_runtime_enablement):
(async_message_statement):
* Source/WebKit/Scripts/webkit/messages_unittest.py:
* Source/WebKit/Scripts/webkit/model.py:
(Message.__init__):
* Source/WebKit/Scripts/webkit/parser.py:
(parse):
* Source/WebKit/Scripts/webkit/tests/Makefile:
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::jsValueForArguments):
(IPC::jsValueForReplyArguments):
(IPC::messageArgumentDescriptions):
(IPC::messageReplyArgumentDescriptions):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessageReceiver.cpp:
(WebKit::TestWithEnabledBy::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessageReceiver.cpp: Removed.
* Source/WebKit/Scripts/webkit/tests/TestWithValidator.messages.in: Renamed from Source/WebKit/Scripts/webkit/tests/TestWithEnabledIf.messages.in.
* Source/WebKit/Scripts/webkit/tests/TestWithValidatorMessageReceiver.cpp: Added.
(WebKit::TestWithValidator::didReceiveMessage):
(IPC::jsValueForDecodedMessage<MessageName::TestWithValidator_AlwaysEnabled>):
(IPC::jsValueForDecodedMessage<MessageName::TestWithValidator_EnabledIfPassValidation>):
(IPC::jsValueForDecodedMessage<MessageName::TestWithValidator_EnabledIfSomeFeatureEnabledAndPassValidation>):
* Source/WebKit/Scripts/webkit/tests/TestWithValidatorMessages.h: Renamed from Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessages.h.
(Messages::TestWithValidator::messageReceiverName):
(Messages::TestWithValidator::AlwaysEnabled::name):
(Messages::TestWithValidator::AlwaysEnabled::AlwaysEnabled):
(Messages::TestWithValidator::AlwaysEnabled::arguments):
(Messages::TestWithValidator::EnabledIfPassValidation::name):
(Messages::TestWithValidator::EnabledIfPassValidation::EnabledIfPassValidation):
(Messages::TestWithValidator::EnabledIfPassValidation::arguments):
(Messages::TestWithValidator::EnabledIfSomeFeatureEnabledAndPassValidation::name):
(Messages::TestWithValidator::EnabledIfSomeFeatureEnabledAndPassValidation::EnabledIfSomeFeatureEnabledAndPassValidation):
(Messages::TestWithValidator::EnabledIfSomeFeatureEnabledAndPassValidation::arguments):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.messages.in:

Canonical link: https://commits.webkit.org/287014@main
